### PR TITLE
Fix problem of small radius

### LINKED ISSUE #873: NiftiSpheresMasker crashes when radius is small
When trying this:

```python
from nilearn.datasets import fetch_adhd
from nilearn.input_data import NiftiSpheresMasker

dataset = fetch_adhd(n_subjects=1)
dmn_coords = [(0, -52, 18), (-46, -68, 32), (46, -68, 32), (0, 50, -5)]
masker = NiftiSpheresMasker(seeds=dmn_coords, radius=1)
data = masker.fit_transform(dataset.func[0])
```

The spheres are then empty.

### PR DESCRIPTION
**Pull request by AlexandreAbraham**
Solves the same problem as #880. Fix #873.

Here is my version of the fix. I didn't tested it extensively but, on the toy script below, it works. The code is super ugly because of the need to switch between seed representation (triplet) and coord_transform one (3 lists of coordinates).

I'm convinced that a better to do it exists but I am too tired to figure it out.

Test script:
```python
from nilearn.input_data import NiftiSpheresMasker
import numpy as np
import nibabel


data = np.random.random((3, 3, 3))
mask = np.zeros((3, 3, 3))
mask[1, 1, 1] = 1
affine = np.eye(4) * 1.2
seed = (1.4, 1.4, 1.4)

masker = NiftiSpheresMasker([seed], radius=0.1,
                            mask_img=nibabel.Nifti1Image(mask, affine))
masker.fit_transform(nibabel.Nifti1Image(data, affine))
```

### CODE REVIEW

**Comment by mrahim** [comment]
Nice.
I have a warning when running my example: ` SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient. `

Also, I'm wondering if it is better to check if the seed coordinates are inside the mask before continuing. 
In the current case, the error message is unclear `seed empty`.

**Comment by AlexandreAbraham** [comment]
> I have a warning when running my example: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient. 

Yeah, it can be solved with one line.

> Also, I'm wondering if it is better to check if the seed coordinates are inside the mask before continuing. 
In the current case, the error message is unclear seed empty.

I do it. However, a masked seed is not a showstopper for me: the rest of the sphere can contain unmasked voxels.

**Comment by bthirion** [comment]
You could ass teh script as a new test. Besides this, LGTM.

File: nilearn/input_data/nifti_spheres_masker.py [hunk 1]
```diff
@@ -50,8 +62,13 @@ def _apply_mask_and_get_affinity(seeds, niimg, radius, allow_overlap,
 
     clf = neighbors.NearestNeighbors(radius=radius)
     A = clf.fit(mask_coords).radius_neighbors_graph(seeds)
+    for i, nearest in enumerate(nearests):
+        if nearest is None:
+            continue
+        A[i, nearest] = True
     A = A.tolil()
     # Include selfs
```
Comment (banilo) [inline nilearn/input_data/nifti_spheres_masker.py]: comment meaning may remain obscure

**Comment by AlexandreAbraham** [comment]
I have fixed the Travis failure. However, this failure was due to the fact that scikit learn crashes if we give a mask with only one voxel unmasked to the neighbor graph estimator... There is probably a bug under the hood.

**Comment by GaelVaroquaux** [comment]
If you can isolate the scikit-learn bug, it would be great to report it upstream.

Did you remove the warning?

**Comment by AlexandreAbraham** [comment]
> If you can isolate the scikit-learn bug, it would be great to report it upstream. 

Added on my todo list.

> Did you remove the warning?

Yes.

**Comment by GaelVaroquaux** [comment]
AppVeyor failed, and I don't understand why. Can you "commit --amend" to trigger a rebuild.

**Comment by lesteve** [comment]
You have a few flake8 errors, run `./continuous_integration/flake8_diff.sh` or look at [this Travis log](https://travis-ci.org/nilearn/nilearn/jobs/105446514).

**Comment by GaelVaroquaux** [comment]
Appveyor is systematically failing on this guy. I am not very comfortable merging it.

**Comment by AlexandreAbraham** [comment]
Build triggered again.

**Comment by AlexandreAbraham** [comment]
Now it's becoming funny.

In this build: https://ci.appveyor.com/project/nilearn-ci/nilearn/build/1.0.577
* build failed on Python 2.7
* build successful on Python 3.4

In this one: https://ci.appveyor.com/project/nilearn-ci/nilearn/build/1.0.575
* build successful on Python 2.7
* build cancelled on Python 3.4

So, the tests are not intrinsically broken but it seems to introduce some hazardous segfaults in appveyor... I don't know what to do.

**Comment by GaelVaroquaux** [comment]
Is it the always same test that is segfaulting?

**Comment by lesteve** [comment]
> In this build: https://ci.appveyor.com/project/nilearn-ci/nilearn/build/1.0.577

>    build failed on Python 2.7


The tests actually seem to run fine with the output:

```
Ran 387 tests in 228.284s
 OK (SKIP=1)
```

I have seen that happening from time to time, but it tends to go away if you kick off another AppVeyor build.

> In this one: https://ci.appveyor.com/project/nilearn-ci/nilearn/build/1.0.575

> build cancelled on Python 3.4

No idea why the build was cancelled.

**Comment by aabadie** [comment]
> No idea why the build was cancelled.

New jobs requests can cancel previously started jobs. You probably pushed a new commit to that branch (or another PR was updated in the mean time). Maybe we can change this settings on appveyor.

**Comment by AlexandreAbraham** [comment]
> Is it the always same test that is segfaulting?

There is no segfault in the tests, it's at the end of the tests.

See other people as troubled as we are:
https://github.com/dart-lang/test/issues/356

**Comment by AlexandreAbraham** [comment]
OK. Looking at the history, the problem on AppVeyor seems to happen quite often. I'm going to merge this one but we should investigate this problem at some point.
